### DISCUSSION TITLE
Redesign the models and interactions between them

### DISCUSSION THREAD

**Discussion by VietAnh1010** [joulev/ezkomment#94] [Questions]
After completing writing routes to handler users, we shall implement middlewares to handle sites, pages and comments.

First, what structure should be used to organize data on Firestore: Multiple nested collections, or simple, flat collections? In the current design, we have:
- `sites` collection that stores the information of every site, without any subcollection.
- `pages` collection that stores the information of every page. In each `page` document there is a `comments` subcollection that contains all comments inside that page.

We can identify a comment with its id and the page's id in the current implementation.

Should we change the design to a nested structure, for instance: `users (identified by uid)` &rarr; `sites (identified by siteId)` &rarr; `pages (identified by pageId)` &rarr; `comments (identified by commentId)`. This design will better demonstrate the relations between the entities (a user has multiple sites, a site has multiple pages, etc.) &ndash; but it will be difficult to access a resource (the request will be clustered with redundant information).

Apart from the collection structures, another problem is how to ensure the uniqueness of some fields.

For example, the `name` of all `sites` of a particular `user` must be unique. Firestore does not have any utility to enforce this. We can manually implement it, by querying the documents and checking them before inserting new data into Firestore.

However, this will increase the number of accesses to the database, which is not desirable. Is there any other way to work around this?

**Comment by joulev** [comment]
> Should we change the design to a nested structure, for instance: users (identified by uid) → sites (identified by siteId) → pages (identified by pageId) → comments (identified by commentId). This design will better demonstrate the relations between the entities (a user has multiple sites, a site has multiple pages, etc.) – but it will be difficult to access a resource (the request will be clustered with redundant information).

This approach sounds fine to me. Does Firebase also include data from subcollections when we read a document? If not, use this approach.

> For example, the name of all sites of a particular user must be unique. Firestore does not have any utility to enforce this. We can manually implement it, by querying the documents and checking them before inserting new data into Firestore.
>
> However, this will increase the number of accesses to the database, which is not desirable. Is there any other way to work around this?

I suggest maintaining an array in the user document

```js
id: "<uid>",
name: "John Doe",
...: ...:,
sites: [
  {
    siteId: "abcdef",
    siteName: "hello-world"
  },
  ...
]
```

**Comment by VietAnh1010** [comment]
> This approach sounds fine to me. Does Firebase also include data from subcollections when we read a document? If not, use this approach.

If we change the design, how should the routes be redesigned? One possible design maybe:
- `/api/users/[uid]`
- `/api/users/[uid]/sites/[sitesId]`
- `/api/users/[uid]/sites/[sitesId]/pages/[pageId]`
- `/api/users/[uid]/sites/[sitesId]/pages/[pageId]/comments/[commentId]`

> I suggest maintaining an array in the user document

Sound good, will try it.

**Reply by joulev** [reply]
> If we change the design, how should the routes be redesigned? One possible design maybe:
>
> /api/users/[uid]
> /api/users/[uid]/sites/[sitesId]
> /api/users/[uid]/sites/[sitesId]/pages/[pageId]
> /api/users/[uid]/sites/[sitesId]/pages/[pageId]/comments/[commentId]

I'm fine with this, but I'm fine with `/api/sites/:id` and `/api/pages/:id` too. It doesn't look like there are any technical problems with either. So use whatever you prefer.

**Reply by joulev** [reply]
Also GitHub discussions is shaped like a StackOverflowish Q&A place so this "answer" feels more like a comment to my answer above :)

**Reply by VietAnh1010** [reply]
`/api/sites/:id` will not be sufficient to identify a site (a site exists in a subcollection that belongs to a user document, therefore we will need a reference to a user document before we can access the subcollections).

I don't really like deeply nested paths, so I will stick with the current design at the moment.
